File: packages/tonic-coding-hydration/src/chunker/astGrepChunker.ts
import * as fs from "node:fs";
import * as path from "node:path";

import type { AstGrepChunk } from "./types";

/** Minimal match shape for chunking (matches tonic-ast-hydration matches[]). */
export type ChunkAstMatch = {
  path: string;
  rule_id: string;
  start?: { line?: number; column?: number };
  end?: { line?: number; column?: number };
  meta?: Record<string, unknown>;
};

function lineRange(m: ChunkAstMatch): { lo: number; hi: number } {
  const lo = m.start?.line ?? 1;
  const hi = Math.max(lo, m.end?.line ?? lo);
  return { lo, hi };
}

function mergeRanges(ranges: Array<{ lo: number; hi: number }>, maxSpan: number): Array<{ lo: number; hi: number }> {
  if (ranges.length === 0) {
    return [];
  }
  const sorted = [...ranges].sort((a, b) => a.lo - b.lo || a.hi - b.hi);
  const out: Array<{ lo: number; hi: number }> = [];
  let cur = { ...sorted[0]! };
  for (let i = 1; i < sorted.length; i++) {
    const r = sorted[i]!;
    if (r.lo <= cur.hi + 1) {
      cur.hi = Math.max(cur.hi, r.hi);
    } else {
      out.push(cur);
      cur = { ...r };
    }
  }
  out.push(cur);
  return out.map((r) => {
    const span = r.hi - r.lo + 1;
    if (span <= maxSpan) {
      return r;
    }
    return { lo: r.lo, hi: r.lo + maxSpan - 1 };
  });
}

/**
 * Derive text chunks from ast-grep match line spans for one file.
 */
export function chunkFileAstGrep(repoRoot: string, relPath: string, matches: ChunkAstMatch[]): AstGrepChunk[] {
  const rel = relPath.replace(/\\/g, "/");
  const fileMatches = matches.filter((m) => m.path.replace(/\\/g, "/") === rel);
  if (fileMatches.length === 0) {
    return [];
  }
  const abs = path.join(repoRoot, rel);
  let lines: string[];
  try {
    const raw = fs.readFileSync(abs, "utf8");
    lines = raw.split(/\r?\n/);
  } catch {
    return [];
  }
  const merged = mergeRanges(fileMatches.map((m) => lineRange(m)), 120);
  const out: AstGrepChunk[] = [];
  let idx = 0;
  for (const rng of merged) {
    idx += 1;
    const start_line = rng.lo;
    const end_line = rng.hi;
    const slice = lines.slice(start_line - 1, end_line).join("\n");
    const first = fileMatches.find((m) => {
      const r = lineRange(m);
      return r.lo <= end_line && r.hi >= start_line;
    });
    const metaName = first?.meta && typeof first.meta.name === "string" ? first.meta.name : undefined;
    out.push({
      path: rel,
      start_line,
      end_line,
      text: slice,
      ast_rule_id: first?.rule_id,
      ast_match_id: `${rel}:${start_line}-${end_line}:${idx}`,
      symbol: metaName,
    });
  }
  return out;
}

/** Chunk every file that appears in the ast-grep match list (stable path order). */
export function chunksFromAstArtifact(repoRoot: string, matches: ChunkAstMatch[]): AstGrepChunk[] {
  const byPath = new Map<string, ChunkAstMatch[]>();
  for (const m of matches) {
    const p = m.path.replace(/\\/g, "/");
    const list = byPath.get(p) ?? [];
    list.push(m);
    byPath.set(p, list);
  }
  const chunks: AstGrepChunk[] = [];
  for (const rel of [...byPath.keys()].sort()) {
    chunks.push(...chunkFileAstGrep(repoRoot, rel, byPath.get(rel)!));
  }
  return chunks;
}

File: merge-tonic-lib/tonic/hydration/ast_grep_chunker.py
"""Ast-grep match spans → text chunks (parity with @mergetonic/coding-hydration)."""

from __future__ import annotations

import math
from pathlib import Path
from typing import Any


def _line_range(m: dict[str, Any]) -> tuple[int, int]:
    start = m.get("start") or {}
    end = m.get("end") or {}
    lo = int(start.get("line", 1)) if isinstance(start, dict) else 1
    hi = int(end.get("line", lo)) if isinstance(end, dict) else lo
    return lo, max(lo, hi)


def _merge_ranges(ranges: list[tuple[int, int]], max_span: int = 120) -> list[tuple[int, int]]:
    if not ranges:
        return []
    sorted_r = sorted(ranges, key=lambda x: (x[0], x[1]))
    out: list[tuple[int, int]] = []
    cur_lo, cur_hi = sorted_r[0]
    for lo, hi in sorted_r[1:]:
        if lo <= cur_hi + 1:
            cur_hi = max(cur_hi, hi)
        else:
            out.append((cur_lo, cur_hi))
            cur_lo, cur_hi = lo, hi
    out.append((cur_lo, cur_hi))
    merged: list[tuple[int, int]] = []
    for lo, hi in out:
        span = hi - lo + 1
        if span > max_span:
            merged.append((lo, lo + max_span - 1))
        else:
            merged.append((lo, hi))
    return merged


def chunk_file_ast_grep(repo_root: str, rel_path: str, matches: list[dict[str, Any]]) -> list[dict[str, Any]]:
    rel = rel_path.replace("\\", "/")
    file_matches = [m for m in matches if str(m.get("path", "")).replace("\\", "/") == rel]
    if not file_matches:
        return []
    abs_path = Path(repo_root) / rel
    try:
        raw = abs_path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return []
    lines = raw.splitlines()
    merged = _merge_ranges([_line_range(m) for m in file_matches])
    out: list[dict[str, Any]] = []
    for idx, (start_line, end_line) in enumerate(merged, start=1):
        slice_lines = lines[start_line - 1 : end_line]
        text = "\n".join(slice_lines)
        first = next(
            (m for m in file_matches if _line_range(m)[0] <= end_line and _line_range(m)[1] >= start_line),
            None,
        )
        meta = first.get("meta") if isinstance(first, dict) else None
        sym = meta.get("name") if isinstance(meta, dict) and isinstance(meta.get("name"), str) else None
        rule_id = str(first.get("rule_id", "")) if first else ""
        out.append(
            {
                "path": rel,
                "start_line": start_line,
                "end_line": end_line,
                "text": text,
                "ast_rule_id": rule_id or None,
                "ast_match_id": f"{rel}:{start_line}-{end_line}:{idx}",
                "symbol": sym,
            }
        )
    return out


def chunks_from_ast_artifact(repo_root: str, matches: list[dict[str, Any]]) -> list[dict[str, Any]]:
    by_path: dict[str, list[dict[str, Any]]] = {}
    for m in matches:
        if not isinstance(m, dict):
            continue
        p = str(m.get("path", "")).replace("\\", "/")
        by_path.setdefault(p, []).append(m)
    chunks: list[dict[str, Any]] = []
    for rel in sorted(by_path.keys()):
        chunks.extend(chunk_file_ast_grep(repo_root, rel, by_path[rel]))
    return chunks
